### ISSUE TITLE
Incorrect schema reference when extending a generic controller

### ISSUE THREAD

**Issue by gabordicso** [micronaut-projects/micronaut-openapi#193]
The generated OpenAPI definition .yml contains an incorrect schema reference when extending a generic controller, unless the endpoint method is explicitly overridden in the child class. See the example below. The method uri is specified in the base class method's Get annotation, but the Controller path is specified in the Controller annotation of the child class, so for the endpoint path constructed from the two, I would expect the generated API definition to reference the generic type of the child class, not the base class. Example:

Micronaut version: 1.3.1

TestBaseObject:

`open class TestBaseObject (var str: String)`

TestChildObject:

`class TestChildObject (str: String, var id: Long? = null) : TestBaseObject(str)`

TestBaseService (this is the generic controller to be extended):

```
open class TestBaseService<T: TestBaseObject>
constructor() {
    lateinit var obj: T

    @Get(uri = "testMethodOnlyInBase")
    fun testMethodOnlyInBase(): T {
        return obj
    }
}
```

TestChildService (this is the actual controller to generate API definition for):

```
@Controller("/api/v1/test")
class TestChildService: TestBaseService<TestChildObject>() {
}
```

Expected .yml output:

```
  /api/v1/test/testMethodOnlyInBase:
    get:
      operationId: testMethodOnlyInBase
      parameters: []
      responses:
        default:
          description: testMethodOnlyInBase default response
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/TestChildObject'
```

Actual .yml (difference in last line):

```
  /api/v1/test/testMethodOnlyInBase:
    get:
      operationId: testMethodOnlyInBase
      parameters: []
      responses:
        default:
          description: testMethodOnlyInBase default response
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/TestBaseObject'
```

Link to complete example with additional use cases covered: [https://github.com/gabordicso/public/tree/master/micronaut/openapi](https://github.com/gabordicso/public/tree/master/micronaut/openapi)

**Comment by istvanszoboszlai** [comment]
Hello, is there something blocking with fixing this issue? Do you guys need help?

**Comment by graemerocher** [comment]
@croudet is https://github.com/micronaut-projects/micronaut-core/pull/3069 related?

**Comment by croudet** [comment]
In some way it is. As some classes and mehtods are visited when they don't have the required annotation or when overriden methods are also processed.

The other part of the problem I thin is this call:
https://github.com/micronaut-projects/micronaut-openapi/blob/b0b45ab9230100bd8aa404e61dc9db3ee5a5eb74/openapi/src/main/java/io/micronaut/openapi/visitor/OpenApiControllerVisitor.java#L223
I think it should be `getGenericReturnType`. Idem for parameters.

**Comment by graemerocher** [comment]
I haven't merged micronaut-projects/micronaut-core#3069  yet because I need to verify it doesn't break Micronaut Data

**Comment by croudet** [comment]
```groovy
    void "test controller inheritance with generics - Issue #193"() {
        given: "An API definition"

        when:
        buildBeanDefinition('test.MyBean', '''
package test;

import javax.annotation.Nullable;
import io.micronaut.http.HttpResponse;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Get;
import io.micronaut.http.annotation.Post;
import io.micronaut.http.annotation.Body;
import io.micronaut.http.annotation.QueryValue;
class BaseObject {
    private int a;
    public int getA() {
        return a;
    }
    public void setA(int a) {
        this.a = a;
    }
}
class B extends BaseObject {
    private int b;
    public int getB() {
        return b;
    }
    public void setB(int b) {
        this.b = b;
    }
}
@Controller("/fromTestOperations")
class TestOperations extends BaseTestOperations<B> {
    @Get("/fromBaseNoGet")
    @Override
    public B getFromBaseNoAnnot() {
        return null;
    }
    @Get("/getOnlyFromB")
    public B getOnlyFromB() {
        return null;
    }
    @Override
    @Get("/fromBaseWithAnnotOverrideB")
    public B getFromBaseWithAnnotOverride() {
        return null;
    }
    @Override
    public B getFromBaseWithNoAnnotOverride() {
        return null;
    }
}
abstract class BaseTestOperations<T extends BaseObject> {
    @Get("/fromBaseOnly")
    public T getFromBaseWithAnnot() {
        return null;
    }
    @Post("/postFromBaseOnly{?foo}")
    public HttpResponse<T> postFromBaseWithAnnot(@Body T value, @QueryValue @Nullable String foo) {
        return null;
    }
    public T getFromBaseNoAnnot() {
        return null;
    }
    public T getFromBaseNoAnnotNoOverride() {
        return null;
    }
    @Get("/fromBaseWithAnnotOverride")
    public T getFromBaseWithAnnotOverride() {
        return null;
    }
    @Get("/fromBaseWithNoAnnotOverride")
    public T getFromBaseWithNoAnnotOverride() {
        return null;
    }
}
@javax.inject.Singleton
class MyBean {}
''')
        then: "the state is correct"
        AbstractOpenApiVisitor.testReference != null

        when: "The OpenAPI is retrieved"
        OpenAPI openAPI = AbstractOpenApiVisitor.testReference
        Schema baseSchema = openAPI.components.schemas['BaseObject']
        Schema bSchema = openAPI.components.schemas['B']

        then: "the components are valid"
		openAPI.paths.size() == 6
        baseSchema != null
        bSchema != null

        when:
        Operation fromBaseOnlyOperation = openAPI.paths.get("/fromTestOperations/fromBaseOnly").get

        then:
        fromBaseOnlyOperation
        fromBaseOnlyOperation.responses
        fromBaseOnlyOperation.responses.size() == 1
        fromBaseOnlyOperation.responses."default"
        fromBaseOnlyOperation.responses."default".content
        fromBaseOnlyOperation.responses."default".content."application/json"
        fromBaseOnlyOperation.responses."default".content."application/json".schema
        fromBaseOnlyOperation.responses."default".content."application/json".schema.$ref
        fromBaseOnlyOperation.responses."default".content."application/json".schema.$ref == '#/components/schemas/B'

        when:
        Operation fromBaseNoGetOperation = openAPI.paths.get("/fromTestOperations/fromBaseNoGet").get

        then:
        fromBaseNoGetOperation
        fromBaseNoGetOperation.responses
        fromBaseNoGetOperation.responses.size() == 1
        fromBaseNoGetOperation.responses."default"
        fromBaseNoGetOperation.responses."default".content
        fromBaseNoGetOperation.responses."default".content."application/json"
        fromBaseNoGetOperation.responses."default".content."application/json".schema
        fromBaseNoGetOperation.responses."default".content."application/json".schema.$ref
        fromBaseNoGetOperation.responses."default".content."application/json".schema.$ref == '#/components/schemas/B'
    }
```

Micronaut 1.3.4 + PR
```
visitClass test.TestOperations
visitMethod getFromBaseNoAnnot() - declaringType: test.TestOperations - owningType: test.TestOperations - enclosingElement: test.TestOperations
visitMethod getOnlyFromB() - declaringType: test.TestOperations - owningType: test.TestOperations - enclosingElement: test.TestOperations
visitMethod getFromBaseWithAnnotOverride() - declaringType: test.TestOperations - owningType: test.TestOperations - enclosingElement: test.TestOperations
visitMethod getFromBaseWithNoAnnotOverride() - declaringType: test.TestOperations - owningType: test.TestOperations - enclosingElement: test.TestOperations
visitMethod getFromBaseWithAnnot() - declaringType: test.BaseTestOperations - owningType: test.TestOperations - enclosingElement: test.BaseTestOperations
visitMethod postFromBaseWithAnnot(T,java.lang.String) - declaringType: test.BaseTestOperations - owningType: test.TestOperations - enclosingElement: test.BaseTestOperations
```

Micronaut 1.3.4
```
visitClass test.TestOperations
visitClass test.BaseTestOperations
visitMethod getFromBaseWithAnnot() - declaringType: test.BaseTestOperations - owningType: test.BaseTestOperations - enclosingElement: test.BaseTestOperations
visitMethod postFromBaseWithAnnot(T,java.lang.String) - declaringType: test.BaseTestOperations - owningType: test.BaseTestOperations - enclosingElement: test.BaseTestOperations
visitMethod getFromBaseWithAnnotOverride() - declaringType: test.BaseTestOperations - owningType: test.BaseTestOperations - enclosingElement: test.BaseTestOperations
visitMethod getFromBaseWithNoAnnotOverride() - declaringType: test.BaseTestOperations - owningType: test.BaseTestOperations - enclosingElement: test.BaseTestOperations
visitMethod getFromBaseNoAnnot() - declaringType: test.TestOperations - owningType: test.BaseTestOperations - enclosingElement: test.TestOperations
visitMethod getOnlyFromB() - declaringType: test.TestOperations - owningType: test.BaseTestOperations - enclosingElement: test.TestOperations
visitMethod getFromBaseWithAnnotOverride() - declaringType: test.TestOperations - owningType: test.BaseTestOperations - enclosingElement: test.TestOperations
visitMethod getFromBaseWithNoAnnotOverride() - declaringType: test.TestOperations - owningType: test.BaseTestOperations - enclosingElement: test.TestOperations
```

- base class processed (no `@Controller` anntotation)
- overriden methods processed 
- owningType not correct for `getOnlyFromB`
- owningType not correct for `postFromBaseWithAnnot`
- declaringType and owningType may not be correct for `getFromBaseNoAnnot`, `getFromBaseWithAnnotOverride` and `getFromBaseWithNoAnnotOverride`

Micronaut 1.1.4
```
visitClass test.TestOperations
visitClass test.BaseTestOperations
visitMethod getFromBaseWithAnnot() - declaringType: test.BaseTestOperations - owningType: test.BaseTestOperations - enclosingElement: test.BaseTestOperations
visitMethod postFromBaseWithAnnot(T,java.lang.String) - declaringType: test.BaseTestOperations - owningType: test.BaseTestOperations - enclosingElement: test.BaseTestOperations
visitMethod getFromBaseWithAnnotOverride() - declaringType: test.BaseTestOperations - owningType: test.BaseTestOperations - enclosingElement: test.BaseTestOperations
visitMethod getFromBaseWithNoAnnotOverride() - declaringType: test.BaseTestOperations - owningType: test.BaseTestOperations - enclosingElement: test.BaseTestOperations
visitMethod getFromBaseNoAnnot() - declaringType: test.TestOperations - owningType: test.BaseTestOperations - enclosingElement: test.TestOperations
visitMethod getOnlyFromB() - declaringType: test.TestOperations - owningType: test.BaseTestOperations - enclosingElement: test.TestOperations
visitMethod getFromBaseWithAnnotOverride() - declaringType: test.TestOperations - owningType: test.BaseTestOperations - enclosingElement: test.TestOperations
visitMethod getFromBaseWithNoAnnotOverride() - declaringType: test.TestOperations - owningType: test.BaseTestOperations - enclosingElement: test.TestOperations

micronaut 1.2.11
```
visitClass test.TestOperations
visitClass test.BaseTestOperations
visitMethod getFromBaseWithAnnot() - declaringType: test.BaseTestOperations - owningType: test.BaseTestOperations - enclosingElement: test.BaseTestOperations
visitMethod postFromBaseWithAnnot(T,java.lang.String) - declaringType: test.BaseTestOperations - owningType: test.BaseTestOperations - enclosingElement: test.BaseTestOperations
visitMethod getFromBaseWithAnnotOverride() - declaringType: test.BaseTestOperations - owningType: test.BaseTestOperations - enclosingElement: test.BaseTestOperations
visitMethod getFromBaseWithNoAnnotOverride() - declaringType: test.BaseTestOperations - owningType: test.BaseTestOperations - enclosingElement: test.BaseTestOperations
visitMethod getFromBaseNoAnnot() - declaringType: test.TestOperations - owningType: test.BaseTestOperations - enclosingElement: test.TestOperations
visitMethod getOnlyFromB() - declaringType: test.TestOperations - owningType: test.BaseTestOperations - enclosingElement: test.TestOperations
visitMethod getFromBaseWithAnnotOverride() - declaringType: test.TestOperations - owningType: test.BaseTestOperations - enclosingElement: test.TestOperations
visitMethod getFromBaseWithNoAnnotOverride() - declaringType: test.TestOperations - owningType: test.BaseTestOperations - enclosingElement: test.TestOperations
```

**Comment by istvanszoboszlai** [comment]
Hello, is there any workaround right now to have correct openapi yaml generated when using generic controllers, or is the only solution to wait for micronaut-projects/micronaut-core#3069 to get merged?

**Comment by croudet** [comment]
Have you tried with latest version: micronaut-core-1.3.6 and micronaut-openapi:1.5.1

**Comment by istvanszoboszlai** [comment]
Oh, great, thanks for the question. The example in the original question does work with micronaut-core-1.3.6 and micronaut-openapi:1.5.1.

Our code, however,  where we use micronaut openapi does now work yet. In our case the indirection is even more complicated, so I will take a look what goes wrong there, and post it If I find the cause.

It does look promising!

**Comment by istvanszoboszlai** [comment]
Okay I have found the problem.
In our project, we have `Single<List<T>>` return types of some methods. In that case the Parent object will be used in the reference. `List<T>` and `Single<T>` both work. I guess the generic type of a generic type is "too much" right now.

I also noticed that if a controller method has a generic paremeter like:
```
  @Put
  fun sumMethod(myChildObject: E)
```
the concrete type of E will also not be inferred.

@croudet , would it help If I would add these use cases to the linked github example project, or would it be better to put these into your test cases in form of a PR?
